### ISSUE TITLE
Database refresh doesn't work within Unit testing

### ISSUE THREAD

**Issue by mmochetti** [laravel/lumen-framework#108]
Every time we try to reset the database (migrate:refresh and seed) from the testing box (phpunit) we get this exception:

exception 'ReflectionException' with message 'Class config does not exist' in /www/vendor/illuminate/container/Container.php:776

Replication steps:
  1) In your **TestCase** base class (commonly located in /tests folder) override method setUp and define it like this:
```php 
public function setUp()
{
     parent::setUp();
     $this->artisan('migrate:refresh');
     $this->seed();
}
```

 2) Run tests from the console using **phpunit** command.

**Note**: this issue also happens if you try to wrap the test within a database transaction and roll it back at the end of the test:
```php 
public function setUp()
{
    parent::setUp();
    DB::beginTransaction();
}

public function tearDown()
{
     DB::rollback();
     parent::tearDown();
}
```

**Comment by zot24** [comment]
+1

**Comment by dwightwatson** [comment]
I'm not sure this is specific to just database refreshing. I have a test suite that does not run any migrations on the database. The first functional test will pass but any subsequent functional tests will fail with a 500 and the very same exception.

    Class config does not exist

It seems as though the application is not refreshing correctly between tests. You can get around this in the meantime by adding a `setUp()` method to your application's `TestCase`:

    public function setUp()
    {
        parent::setUp();

        $this->createApplication();
    }

It's curious that this works, under the hood Lumen is calling `$this->app = $this->createApplication()` and I've yet to work out why the method needs to be called twice in order for the suite to work.

There is a discussion about this issue [on the Laravel.io forums](http://laravel.io/forum/04-22-2015-lumen-problem-with-migrations-in-unit-tests-class-config-does-not-exist). Additionally, issue #96 appears to be related.

This indeed appears to be a bug with Lumen.

**Comment by mmochetti** [comment]
Exactly what it happened to me. The first time it worked, next subsequent test runs wont. 
What I did to work around this was to create an artisan command that "test:unit" that would refresh --seed the datbase before running the tetsts each time (this is mainly because the refresh --seed has the same issue if you execute it between tests).

**Comment by jaduquem** [comment]
If you have a db call like this in any of your seeders:

```php
DB::table('users')->delete();
```

Then the issue is that `DB` is not being resolved correctly, that is, is not being resolved from the main app container, but rather from a generic one (I think).

Anyway, as a temporary fix you could write those sentences like this:
```php
$this->container['db']->table('user')->truncate();
```

In a more general form, change this:

```php
DB::
```

for this:

```php
$this->container['db']->
```

**Comment by KarimGeiger** [comment]
In addition, you may have a look at the issue https://github.com/laravel/framework/issues/8547 - seems like the same problem.

**Comment by tehs0ap** [comment]
I've been running into these strange issues all day trying to get testing setup.

The symptoms have varied:
* Getting 'Class config does not exist' exceptions when trying to `tearDown`.
* Getting 'SQLSTATE[HY000]: General error: 1 no such table' when trying to use Artisan facade to migrate / seed in `createApplication`
* Using the session helper in a `setUp` and not having it set correctly when running more then one test
* Getting validation errors using the exists, again on the nth test.

I have facades turned on and I'm confident using them is causing the issues, specifically testing code that is using them.
I think the problem comes from the fact that the application is recreated for every single test and facades are implemented as singletons. These singletons come from the instance of the application they were first called from.

From Illuminate/Support/Facades/Facade.php
``` php
	/**
	 * Resolve the facade root instance from the container.
	 *
	 * @param  string  $name
	 * @return mixed
	 */
	protected static function resolveFacadeInstance($name)
	{
		if (is_object($name)) return $name;

		if (isset(static::$resolvedInstance[$name]))
		{
			return static::$resolvedInstance[$name];
		}

		return static::$resolvedInstance[$name] = static::$app[$name];
	}
```

I haven't really dug much deeper than this yet, but it seems like information specific to the instance of the application must be stored in the singleton facade instance. Maybe things such as references for the in memory db and session.

What confirmed my suspicions but clearly breaks facades as singletons was to have `resolveFacadeInstance()` return `static::$app[$name];`.

Following that train of thought I think you can also override the application creation to get the same effect when unit testing.
```php
public function createApplication()
{
    $app = require __DIR__.'/../bootstrap/app.php';

    \Illuminate\Support\Facades\Facade::clearResolvedInstances();    

    return $app;
}
```

Hopefully this can help pinpoint the issue and help find a better solution.

Side note: I also noticed that `$this->app` does not carry over from test to test even within the same test case file. Not sure if this is a PHPUnit quirk, or I'm missing something else entirely but it make the check in Laravel\Lumen\Testing\TestCase `setUp` pointless.

**Comment by sebastiaanluca** [comment]
Got a similar issue where it throws a `Class … already declared` exception when (only) using `refresh`.

**Comment by GrahamCampbell** [comment]
This is a phpunit bug.

**Comment by taylorotwell** [comment]
I can't duplicate this at all. Do you have withEloquent() uncommented in your bootstrap/app.php file?

**Comment by jaduquem** [comment]
I do have withEloquent() uncommented.

**Comment by zot24** [comment]
Yes, I have `withEloquent()` uncommented on my `app.php` file

**Comment by taylorotwell** [comment]
Hmm, I just can't recreate it on a fresh Lumen install.

**Comment by dwightwatson** [comment]
I found all my issues with Lumen testing (which I described above) were resolved in Lumen 5.1.

**Comment by tehs0ap** [comment]
I think this issue was fixed in c5d599b086f20a6c0db69ba6ff1f00999ef68654

**Comment by MagentoExp** [comment]
> 
> 
> Every time we try to reset the database (migrate:refresh and seed) from the testing box (phpunit) we get this exception:
> 
> exception 'ReflectionException' with message 'Class config does not exist' in /www/vendor/illuminate/container/Container.php:776
> 
> Replication steps:
> 
>     1. In your **TestCase** base class (commonly located in /tests folder) override method setUp and define it like this:
> 
> 
> ```
> public function setUp()
> {
>      parent::setUp();
>      $this->artisan('migrate:refresh');
>      $this->seed();
> }
> ```
> 
>     1. Run tests from the console using **phpunit** command.
> 
> 
> **Note**: this issue also happens if you try to wrap the test within a database transaction and roll it back at the end of the test:
> 
> ```
> public function setUp()
> {
>     parent::setUp();
>     DB::beginTransaction();
> }
> 
> public function tearDown()
> {
>      DB::rollback();
>      parent::tearDown();
> }
> ```

Below it is how I solved it , Laravel 5.7 and Phpunit -6.5

ExampleTest.php
```

use DB;
use Tests\TestCase;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Support\Facades\Session;
use Illuminate\Foundation\Testing\DatabaseMigrations;
use Illuminate\Foundation\Testing\DatabaseTransactions;
use App\Model;
use Faker\Factory;

class ExampleTest extends TestCase {

    use DatabaseTransactions;

    /**
     * A basic test example.
     *
     * @return void
     */
    public function testBasicTest() {
}
}

TestCase.php


```
namespace Tests;

use Illuminate\Foundation\Testing\TestCase as BaseTestCase;

abstract class TestCase extends BaseTestCase {

    use CreatesApplication;

    protected function setUp() {
        parent::setUp();
        \DB::beginTransaction();
        $this->artisan('migrate:refresh');
        $this->artisan('db:seed');
    }

    protected function tearDown() {
        \DB::rollback();
        parent::tearDown();
    }
}
```

```
